drivers/spi/spi_sy1xx.c

```diff
@@ -0,0 +1,521 @@
+
+	if (request->data_len == 0) {
+		return -2;
+	}
+
+	if (request->data_len > DEVICE_MAX_BUFFER_SIZE) {
+		return -3;
+	}
+
+	/* stop any prior transmission */
+	SY1XX_UDMA_WAIT_FOR_FINISHED_TX(config->base);
+	SY1XX_UDMA_WAIT_FOR_FINISHED_RX(config->base);
+
+	uint8_t *buf = &data->write[0];
+
+	uint32_t count = 0;
+
+	/* start writing */
+	if (request->data_len > 0) {
+		count = 0;
+
+		/* cfg */
+		buf[count++] = SPI_CMD_CFG_3(config->div);
+		buf[count++] = SPI_CMD_CFG_2(config->cpol, config->cpha);
+		buf[count++] = SPI_CMD_CFG_1();
+		buf[count++] = SPI_CMD_CFG_0();
+
+		/* start with selecting chip-select */
+		buf[count++] = SPI_CMD_SOT_3(request->cs);
+		buf[count++] = SPI_CMD_SOT_2();
+		buf[count++] = SPI_CMD_SOT_1();
+		buf[count++] = SPI_CMD_SOT_0();
+
+		if (request->addr_width > 0) {
+			/* set address if configured */
+			buf[count++] = SPI_CMD_SEND_ADDR3();
+			buf[count++] = SPI_CMD_SEND_ADDR2();
+			buf[count++] = SPI_CMD_SEND_ADDR1(request->addr_width);
+			buf[count++] = SPI_CMD_SEND_ADDR0(config->quad_spi);
+
+			/* based on width, shift */
+			uint32_t addr;
+
+			if (request->addr_width <= 8) {
+				addr = request->addr << 24;
+			} else if (request->addr_width <= 16) {
+				addr = request->addr << 16;
+			} else if (request->addr_width <= 24) {
+				addr = request->addr << 8;
+			} else {
+				addr = request->addr;
+			}
+
+			/* addr is always uint32 */
+			buf[count++] = UINT32_BYTE3(addr);
+			buf[count++] = UINT32_BYTE2(addr);
+			buf[count++] = UINT32_BYTE1(addr);
+			buf[count++] = UINT32_BYTE0(addr);
+		}
+
+		/* expected data config and spi transfer type */
+		buf[count++] = SPI_CMD_SEND_DATA3(request->data_len * 8);
+		buf[count++] = SPI_CMD_SEND_DATA2(request->data_len * 8);
+		buf[count++] = SPI_CMD_SEND_DATA1();
+		buf[count++] = SPI_CMD_SEND_DATA0(config->quad_spi);
+
+		/* data; we need to fill multiple of 32bit size */
+		uint32_t wordCount = ROUND_UP_DIV_4(request->data_len);
+
+		for (uint32_t i = 0; i < request->data_len; i++) {
+			buf[count + i] = request->data[i];
+		}
+		count += (wordCount * 4);
+
+		/* end of transmission */
+		buf[count++] = SPI_CMD_EOT3(0);
+		buf[count++] = SPI_CMD_EOT2();
+		buf[count++] = SPI_CMD_EOT1();
+		buf[count++] = SPI_CMD_EOT0();
+
+		SY1XX_UDMA_START_TX(config->base, (uint32_t)buf, count, 0);
+
+		SY1XX_UDMA_WAIT_FOR_FINISHED_TX(config->base);
+
+		int32_t xfer_count = request->data_len - SY1XX_UDMA_GET_REMAINING_RX(config->base);
+
+		if ((xfer_count > 0) && (xfer_count <= request->data_len)) {
+			return xfer_count;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @return:
+ *  - < 0: Error
+ *  - = 0: No bytes received
+ *  - > 0: Number of bytes received
+ */
+int32_t sy1xx_spi_read(const struct device *dev, struct sy1xx_spi_xfer *request)
+{
+
+	struct sy1xx_spi_config *config = (struct sy1xx_spi_config *)dev->config;
+	struct sy1xx_spi_data *data = (struct sy1xx_spi_data *)dev->data;
+
+	if (request->data_len == 0) {
+		return -2;
```

Comment:
looks kinda random, why not use errno values?

Reply:
for sure. I changed that.